File: LeetCode/0404-sum-of-left-leaves/0404-sum-of-left-leaves.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int solve(TreeNode root, boolean isLeft) {
        if (root == null) {
            return 0;
        }
        if (root.left == null && root.right == null) {
            if (isLeft) {
                return root.val;
            }
        }
        return solve(root.left, true) + solve(root.right, false);
    }
    public int sumOfLeftLeaves(TreeNode root) {
        return solve(root, false);
    }
}

// class Solution {
//     public int solve(TreeNode root, TreeNode parent) {
//         if (root == null) {
//             return 0;
//         }
//         if (root.left == null && root.right == null) {
//             if (parent != null && parent.left == root) {
//                 return root.val;
//             }
//         }
//         return solve(root.left, root) + solve(root.right, root);
//     }
//     public int sumOfLeftLeaves(TreeNode root) {
//         return solve(root, null);
//     }
// }

File: LeetCode/0404-sum-of-left-leaves/0404-sum-of-left-leaves.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int solve(TreeNode* root, bool isLeft) {
        if (root == nullptr) {
            return 0;
        }
        if (root->left == nullptr && root->right == nullptr) {
            if (isLeft) {
                return root->val;
            }
        }
        return solve(root->left, true) + solve(root->right, false);
    }
    int sumOfLeftLeaves(TreeNode* root) {
        return solve(root, false);
    }
};

// class Solution {
// public:
//     int solve(TreeNode* root, TreeNode* parent) {
//         if (root == nullptr) {
//             return 0;
//         }
//         if (root->left == nullptr && root->right == nullptr) {
//             if (parent != nullptr && parent->left == root) {
//                 return root->val;
//             }
//         }
//         return solve(root->left, root) + solve(root->right, root);
//     }
//     int sumOfLeftLeaves(TreeNode* root) {
//         return solve(root, nullptr);
//     }
// };